### DISCUSSION TITLE
Looking to have a discussion with maintainers about PEP 501

### DISCUSSION THREAD

**Discussion by nhumrich** [sqlalchemy/sqlalchemy#6790] [Ideas]
Hello SQLAlchemy maintainers!

I love sqlalchemy and all you have done for the python community. 

I am working on trying to PEP 501 revisited. If you don't know, PEP 501 is essentially a syntax like f-strings, that allow you defer evaluation of the string. The main purpose of the PEP, is so that downstream libraries, such as SQLAlchemy can render user-inputed data from the string, however they want. Or, in the case of SQL, render data safely. 

For example, SQLAlchemy already has the option of `text`, which allows a developer to bypass the ORM and specificy their own queries. I know you know this, but for the purpose of making sure everyone is on the same page, this looks like this:
```
value = "Unsafe text'  && 1=1 --"
sa.text('SELECT * FROM mytable where name=:name').params(name=value)
```

Sqlalchemy replaces the value with a parameter in SQL, so that no sql injection happens. 

This has two problems however:
1. Now that f-strings are so common, most people will just use f-string, and pass in the value directly instead of writing `:name`.
2. SQLAlchemy is parsing the string for `:` which is a problem if the string actually needs a `:`, for example, when casting in data types, such as `data::text`. If you want to do this, you have to escape all colons. 

PEP 501 would instead allow a developer to type:
```
sql(i'SELECT * FROM mytable where name={value}')
```

and then sqlalchemy could properly escape the value, because it knows its being passed in as a dynamic value, and not part of the literal string. 

Here is a possible implementation (based on my current draft of 501, not as it currently stands):
```
    def sql(template):
        params = []
        final_values = {}
        for i (v, s) in enumerate(template.values):
            sql_key = f'value_{i}'
            params.append(f':{sql_key}')
            final_values[sql_key] = v
        sql_string = ''.join(f'{literal}{value}' for literal, value in zip(template.literals, params))
        return sqlalchemy.text(sql_string).params(**final_values)
```

Of course, you could implement it however you want. Its possible you even have a `sa.textSafe` method or something that is just like `text` but enforces only String Templates (and not normal strings) are passed in. This would allow an even safer way to prevent SQLInjection. Essentially, it's mistake proof, as any raw string would be rejected as unsafe. 

The reason I bring all this up, is that I think having a popular library such as SQLAlchemy commited to using the new python feature if and when it gets accepted, would go a long way in getting the PEP accepted. 
So what do you say? Would SQLAlchemy benefit from this change? Would you be willing to commit to using it (can I put your commitment in the PEP?)
If so, in what ways do you think it will benefit SQLAlchemy?

Also, for what it is worth, I would love your feedback on what features/changes you would add/make to SQLAlchemy with this python feature, but I am happy to help you implement it into SQLAlchemy itself, if you would prefer. (Assuming of course, the PEP is accepted). And of course, if you are interested, I will keep you updated on the PEP as it progresses.

**Comment by zzzeek** [comment]
i will share my thoughts on this soon.

**Reply by zzzeek** [reply]
OK, just to go quickly.   There is a good argument for SQLAlchemy to make use of this feature if it were implemented in Python, however the two reasons offered are not them.     re: "people will just use F strings now", I don't really think so, folks either are string concatenating / templating (like with %s) because they don't know what they're doing (and in my experience, nobody is doing that with text()), or they aren't.  Adding * yet another * syntax isn't going to help people that are already using the inappropriate syntax.  SQLAlchemy has deprecated sending raw strings to `connection.execute()` without using `text()`, the purpose of which is, to get the user to go and learn about `text()`, read the docs, and know that they should make sure it's parametrized.     

re: "sqlalchemy is parsing the colon and needs an escape sequence", well sure, there are also SQL syntaxes that use curly braces and there is still an escape format needed, so there's no "escape from escapes" :)    Of course it's nice that Python will provide the parser for us, that's nice.

here's the reason you *should* use, which is that it's more succinct:

    stmt = text('SELECT * FROM mytable where name=:name').params(name=value)

    stmt = itext(i"SELECT * FROM mytable WHERE name={value}")

it's less verbose and does the job for a simple case, so that's great.

So all of that said, it's important to note that `text()` in SQLAlchemy is not at all the ideal way to use SQLAlchemy.  It's of course first class API but is second-class functionality-wise, as it offers no composability, or platform agnosticism, or intelligent handling of expression or result set mechanics without adding additional directives for which the `i""` format would not solve.     

so on our end, this would benefit SQLAlchemy by giving us a slightly nicer syntax for the already non-first-class use case of using text(), but that's about it.

There's certainly no problem I can see with SQLAlchemy including an `itext()` construct should an acceptable form of pep501 get accepted but I would not feel comfortable with the term "commit".    SQLAlchemy's construction system is not textually oriented so this would be a nice to have for edge cases but wouldn't be a fundamental change.
